Lib/json/encoder.py

```diff
@@ -220,42 +225,46 @@ def iterencode(self, o, _one_shot=False):
         else:
             _encoder = encode_basestring
 
-        def floatstr(o, allow_nan=self.allow_nan,
-                _repr=float.__repr__, _inf=INFINITY, _neginf=-INFINITY):
-            # Check for specials.  Note that this type of test is processor
-            # and/or platform-specific, so do tests which don't depend on the
-            # internals.
-
-            if o != o:
-                text = 'NaN'
-            elif o == _inf:
-                text = 'Infinity'
-            elif o == _neginf:
-                text = '-Infinity'
-            else:
-                return _repr(o)
-
-            if not allow_nan:
-                raise ValueError(
-                    "Out of range float values are not JSON compliant: " +
-                    repr(o))
-
-            return text
-
 
         if (_one_shot and c_make_encoder is not None
                 and self.indent is None):
             _iterencode = c_make_encoder(
-                markers, self.default, _encoder, self.indent,
+                markers, self.default, _encoder, self.indent, self.encode_float,
                 self.key_separator, self.item_separator, self.sort_keys,
                 self.skipkeys, self.allow_nan)
         else:
+            if not self.encode_float:
+                def floatstr(o, allow_nan=self.allow_nan,
+                        _repr=float.__repr__, _inf=INFINITY, _neginf=-INFINITY):
+                    # Check for specials.  Note that this type of test is processor
+                    # and/or platform-specific, so do tests which don't depend on the
+                    # internals.
+
+                    if o != o:
+                        text = 'NaN'
+                    elif o == _inf:
+                        text = 'Infinity'
+                    elif o == _neginf:
+                        text = '-Infinity'
+                    else:
+                        return _repr(o)
+
+                    if not allow_nan:
+                        raise ValueError(
+                            "Out of range float values are not JSON compliant: " +
+                            repr(o))
+
+                    return text
```

Comment:
Minor comment: why move the floatstr function definition inline here?  it could stay as a regular method in the class, then the line below would be `self.encode_float = self.floatstr` and still work.